runtime/interpreter/interpreter.go

```diff
@@ -3817,8 +3837,37 @@ func (interpreter *Interpreter) checkTypeLoading(staticType StaticType) (typeErr
 		}
 	}()
 
-	// Here it is only interested in whether the type can be properly loaded.
-	_, typeError = interpreter.ConvertStaticToSemaType(staticType)
+	// Here, the value at the path could be either:
+	//	1) The actual stored value (storage path)
+	//	2) A link to the value at the storage (private/public paths)
+	//
+	// Therefore, try to find the final path, and try loading the value.
+
+	// However, borrow type is not statically known.
+	// So take the borrow type from the value itself
+
+	var borrowType StaticType
+	if _, ok := value.(LinkValue); ok {
+		// Link values always have a `CapabilityStaticType` static type.
+		borrowType = staticType.(CapabilityStaticType).BorrowType
+	} else {
+		// Reference type with value's type (i.e. `staticType`) as both the borrow type and the referenced type.
+		borrowType = NewReferenceStaticType(interpreter, false, staticType, staticType)
+	}
+
+	var semaType sema.Type
+	semaType, valueError = interpreter.ConvertStaticToSemaType(borrowType)
+	if valueError != nil {
+		return valueError
+	}
+
+	// This is guaranteed to be a reference type, because `borrowType` is always a reference.
+	referenceType, ok := semaType.(*sema.ReferenceType)
+	if !ok {
+		panic(errors.NewUnreachableError())
+	}
+
+	_, valueError = interpreter.checkValueAtPath(address, path, referenceType, EmptyLocationRange)
```

Comment:
Maybe pass in the location range to `Interpreter.checkValue` and pass it along here (if available at call-site)

Reply:
added e5694e7d6d5549f4653e906c78e32d68128eba18